### PR TITLE
buildpack-deps: Update hera and fix some issues.

### PR DESCRIPTION
**Pull request by aarlt**
Replaced by #10425.

This PR was intended to just update hera to `v0.3.1`. Doing this some unrelated issues where identified & fixed:
- ~broken symlink `scripts/ci/buildpack-deps_test_emscripten.sh`: link was pointing to `../../scripts/travis-emscripten/build_emscripten.sh`~ (see #10338)
- due to a security vulnerability `set-env` got deprecated, update `scripts/ci/docker_upgrade.sh` to use [Environment Files](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-commands-for-github-actions#environment-files)

### CODE REVIEW

File: scripts/ci/docker_upgrade.sh
Comment (aarlt) [inline scripts/ci/docker_upgrade.sh]: Due to a [security vulnerability](https://github.com/actions/toolkit/security/advisories/GHSA-mfwh-5m23-j46w) `set-env` got deprecated.  As a new mechanism [Environment Files](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-commands-for-github-actions#environment-files) where introduced.

File: scripts/docker/buildpack-deps/Dockerfile.emscripten
Comment (aarlt) [inline scripts/docker/buildpack-deps/Dockerfile.emscripten:32]: This is unrelated to updating hera. But I updated the version here to verify that `scripts/ci/buildpack-deps_test_emscripten.sh` is now pointing to the correct file and that it can be executed successfully.

**Comment by axic** [comment]
Should be rebased after #10338

**Review by axic — DISMISSED** [review]
Looks okay to me, but someone who has worked on the docker images should verify, such as @ekpyron and @cameel.

**Review by cameel — DISMISSED** [review]
Looks fine to me but you should update #10339 and see it tests pass with the new image IDs there before merging this. They did pass before the rebase so I'm accepting anyway.

Also, looks like we're still using hera 0.3.0 in two places:
- `osx_install_dependencies.sh`
- `test/Common.h`

I think we should update them too, though that could be a separate PR.

**Comment by aarlt** [comment]
@axic @cameel hmm.. I think we should disable `HERA_DEBUGGING` for the buildpacks - its a bit too noisy.. e.g. https://app.circleci.com/pipelines/github/ethereum/solidity/10535/workflows/b92fd221-4380-45af-ba1c-1b0520532c75/jobs/513547/parallel-runs/1/steps/1-103
maybe we should compile a debug and a non-debug version, but in the ci the non-debug version should be used by default.. for now I just disabled  `HERA_DEBUGGING` for the buildpacks

**Comment by leonardoalt** [comment]
@aarlt @cameel is the debug disabling consensus?

**Comment by leonardoalt** [comment]
Ok, back here as pointed by @cameel .
@aarlt why does the debug version print all that stuff? Are debug and verbosity not two different things?

File: scripts/docker/buildpack-deps/Dockerfile.ubuntu1604.clang.ossfuzz
Comment (cameel) [inline scripts/docker/buildpack-deps/Dockerfile.ubuntu1604.clang.ossfuzz:111]: @axic Is there no way to switch this on and off a runtime? I think it would be most useful as an option you could pass to `soltest`/`isoltest` to get extra output when you need it (and when you know you're running just a few tests so you don't get flooded with output).

Comment (cameel) [inline scripts/docker/buildpack-deps/Dockerfile.ubuntu1604.clang.ossfuzz:111]: Adding `-DHERA_DEBUGGING=OFF` is completely unrelated to the update so I think it should be in a separate commit.

**Comment by cameel** [comment]
> I think we should disable HERA_DEBUGGING for the buildpacks - its a bit too noisy

I agree. This output is completely useless in this situation. I think it should not print anything by default just like `evmone` does not.

> @aarlt why does the debug version print all that stuff? Are debug and verbosity not two different things?

Yeah, I also think this should be a runtime switch tied to verbosity, not a compile-time option. But looks like this it how hera works right now.

File: scripts/docker/buildpack-deps/Dockerfile.ubuntu1604.clang.ossfuzz
Comment (axic) [reply scripts/docker/buildpack-deps/Dockerfile.ubuntu1604.clang.ossfuzz:111]: No, it is compile time only. There was a big PR to implement it runtime, but it was abandoned. At the time we were more interested in speed than debuggability.

Comment (cameel) [reply scripts/docker/buildpack-deps/Dockerfile.ubuntu1604.clang.ossfuzz:111]: In that case I think it's better to have it disabled by default. I could be very useful when debugging locally but not really in CI where we'll get debug output for all tests in the repo in one big lump :)

File: scripts/docker/buildpack-deps/Dockerfile.ubuntu1804
Comment (axic) [inline scripts/docker/buildpack-deps/Dockerfile.ubuntu1804:94]: Why isn't the release build good enough? It has debugging off by default.

File: scripts/docker/buildpack-deps/Dockerfile.ubuntu1804
Comment (axic) [inline scripts/docker/buildpack-deps/Dockerfile.ubuntu1804:99]: Why isn't the release build good enough? It has debugging off by default.

File: scripts/docker/buildpack-deps/Dockerfile.ubuntu1604.clang.ossfuzz
Comment (axic) [inline scripts/docker/buildpack-deps/Dockerfile.ubuntu1604.clang.ossfuzz:111]: Debugging is off by default, so no need to specify this?

Comment (aarlt) [reply scripts/docker/buildpack-deps/Dockerfile.ubuntu1604.clang.ossfuzz:111]: I think `HERA_DEBUGGING` is by default on. See https://github.com/ewasm/hera/blob/65efd8c9b6ddaef518d55d42e49c6c4329a3ad0c/src/CMakeLists.txt#L31

File: scripts/docker/buildpack-deps/Dockerfile.ubuntu1804
Comment (aarlt) [reply scripts/docker/buildpack-deps/Dockerfile.ubuntu1804:99]: Debugging is enabled by default.

Comment (axic) [reply scripts/docker/buildpack-deps/Dockerfile.ubuntu1804:99]: Oh, indeed it is, that is a mistake upstream: https://github.com/ewasm/hera/blob/a772fafd7507fa0612837a8c3b1fdb6fd3f2674b/src/CMakeLists.txt#L31

It is not supposed to be on by default.

Comment (axic) [reply scripts/docker/buildpack-deps/Dockerfile.ubuntu1804:99]: Should I make a 0.3.2 release with that disabled, that may be better because then even the `install_dependencies` script will work as intended.

Comment (aarlt) [reply scripts/docker/buildpack-deps/Dockerfile.ubuntu1804:99]: Ah, great. Yes, I would say it would make sense to create a new release.

File: scripts/docker/buildpack-deps/Dockerfile.ubuntu1804
Comment (cameel) [inline scripts/docker/buildpack-deps/Dockerfile.ubuntu1804:94]: I'd suggest using `--depth 1` in clone commands as a general rule unless there's a need to have all history available.

**Review by cameel — COMMENTED** [review]
I think the PR is fine now. Unless we're waiting for 0.3.2 to avoid the custom build on Ubuntu 18.0?

**Comment by axic** [comment]
Pushed 0.3.2, but will need to wait for the CI to finish the release binaries: https://github.com/ewasm/hera/releases/tag/v0.3.2

**Pull request closed without merging**
